course/utils.py

```diff
@@ -1241,94 +1244,114 @@ def __call__(self, *args, **kwargs):
 
 # {{{ ipynb utilities
 
+def get_default_ipynb_render_config():
+    # type: () -> Config
+    c = Config()
+    c.ExecutePreprocessor.enabled = False
+    c.CSSHTMLHeaderPreprocessor.enabled = False
+    c.HighlightMagicsPreprocessor.enabled = False
+    return c
+
+
+def render_notebook_from_source(
+        ipynb_source, indices=None,
+        clear_output=False, clear_markdown=False, config=None,
+        template_file=None, config_callback=None, **kwargs):
+    # type: (Text, Optional[Any], Optional[bool], Optional[bool], Optional[Config], Optional[Text], Optional[Callable], **Any) -> Text  # noqa
+    """
+    Get HTML format of ipython notebook so as to be rendered in RELATE flow
+    pages.
+    :param ipynb_source: the :class:`text` read from a ipython notebook.
+    :param indices: a :class:`list` instance, 0-based indices of notebook cells
+    which are expected to be rendered.
+    :param clear_output: a :class:`bool` instance, indicating whether existing
+    execution output of code cells should be removed.
+    :param clear_markdown: a :class:`bool` instance, indicating whether markdown
+    cells will be ignored..
+    :param config: a :class:`traitlets.config.loader.Config` instance.
+    :param config_callback: a function which further handles `config` .
+    :param template_file: :class:`text`, the name of the template file used. If not
+    specified, built-in basic template will be used.
+    :return: :class:`text`
+    """
+    import nbformat
+    from nbformat.reader import parse_json
+    nb_source_dict = parse_json(ipynb_source)
+
+    if indices:
+        nb_source_dict.update(
+            {"cells": [nb_source_dict["cells"][idx] for idx in indices]})
+
+    if clear_markdown:
+        nb_source_dict.update(
+            {"cells": [cell for cell in nb_source_dict["cells"]
+                       if cell['cell_type'] != "markdown"]})
+
+    nb_source_dict.update({"cells": nb_source_dict["cells"]})
+
+    import json
+    ipynb_source = json.dumps(nb_source_dict)
+    notebook = nbformat.reads(ipynb_source, as_version=4)
+
+    if not config:
+        config = get_default_ipynb_render_config()
+
+    # This is to prevent execution of arbitrary code from note book
+    if clear_output:
+        config.ClearOutputPreprocessor.enabled = True
+
+    if config_callback:
+        config = config_callback(config)
+
+    template_file = template_file or "basic"
+
+    from nbconvert import HTMLExporter
+    html_exporter = HTMLExporter(
+        config=config,
+        template_file=template_file
+    )
+
+    (body, resources) = html_exporter.from_notebook_node(notebook)
+
+    return "<div class='relate-notebook-container'>%s</div>" % body
```

Comment:
This is hard to review, as the code was moved. What was changed in addition to the move?

Reply:
I guess `config_callback` was added. Anything else?

Reply:
> I guess config_callback was added. Anything else?

Yes. Because we need a different template (built-in `basic.tpl` template) as compared to what we do when we call the `render_notebook_cells` method, which uses the [`nbconvert_template.tpl`](https://github.com/inducer/relate/blob/master/course/templates/course/jinja2/nbconvert_template.tpl).

I'll check if there's other changes.